database/sqlite/sqlite_context.c

```diff
@@ -0,0 +1,597 @@
+
+    snprintfz(buf, 1024, "ATTACH DATABASE \"%s/netdata-meta.db\" as meta;", netdata_configured_cache_dir);
+    if(init_database_batch(db_context_meta, DB_CHECK_NONE, 0, list)) return 1;
+
+    if (init_database_batch(db_context_meta, DB_CHECK_NONE, 0, &database_context_config[0]))
+        return 1;
+
+    if (init_database_batch(db_context_meta, DB_CHECK_NONE, 0, &database_context_cleanup[0]))
+        return 1;
+
+    return 0;
+}
+
+/*
+ * Close the sqlite database
+ */
+
+void sql_close_context_database(void)
+{
+    int rc;
+    if (unlikely(!db_context_meta))
+        return;
+
+    info("Closing context SQLite database");
+
+    rc = sqlite3_close_v2(db_context_meta);
+    if (unlikely(rc != SQLITE_OK))
+        error_report("Error %d while closing the context SQLite database, %s", rc, sqlite3_errstr(rc));
+    return;
+}
+
+static int bind_text_null(sqlite3_stmt *res, int position, const char *text)
+{
+    if (likely(text))
+        return sqlite3_bind_text(res, position, text, -1, SQLITE_STATIC);
+    return sqlite3_bind_null(res, position);
+}
+
+//
+// Fetching data
+//
+#define CTX_GET_CHART_LIST  "SELECT c.chart_id, c.type||'.'||c.id, c.name, c.context, c.title, c.unit, c.priority, c.update_every, c.chart_type, c.family FROM meta.chart c " \
+        "WHERE c.host_id IN (SELECT h.host_id FROM meta.host h " \
+        "WHERE UNLIKELY((h.hops = 0 AND @host_id IS NULL)) OR LIKELY((h.host_id = @host_id)));"
+
+void ctx_get_chart_list(uuid_t *host_uuid, void (*dict_cb)(SQL_CHART_DATA *, void *), void *data)
+{
+    int rc;
+    sqlite3_stmt *res = NULL;
+
+    rc = sqlite3_prepare_v2(db_context_meta, CTX_GET_CHART_LIST, -1, &res, 0);
+    if (unlikely(rc != SQLITE_OK)) {
+        error_report("Failed to prepare statement to fetch chart list");
+        return;
+    }
+
+    if (likely(host_uuid))
+        rc = sqlite3_bind_blob(res, 1, *host_uuid, sizeof(*host_uuid), SQLITE_STATIC);
+    else
+        rc = sqlite3_bind_null(res, 1);
+    if (unlikely(rc != SQLITE_OK)) {
+        error_report("Failed to bind host_id to fetch the chart list");
+        goto failed;
+    }
+
+    SQL_CHART_DATA chart_data = { 0 };
+    while (sqlite3_step(res) == SQLITE_ROW) {
+        uuid_copy(chart_data.chart_id, *((uuid_t *)sqlite3_column_blob(res, 0)));
+        chart_data.id = (char *) sqlite3_column_text(res, 1);
+        chart_data.name = (char *) sqlite3_column_text(res, 2);
+        chart_data.context = (char *) sqlite3_column_text(res, 3);
+        chart_data.title = (char *) sqlite3_column_text(res, 4);
+        chart_data.units = (char *) sqlite3_column_text(res, 5);
+        chart_data.priority = sqlite3_column_int(res, 6);
+        chart_data.update_every = sqlite3_column_int(res, 7);
+        chart_data.chart_type = sqlite3_column_int(res, 8);
+        chart_data.family = (char *) sqlite3_column_text(res, 9);
+        dict_cb(&chart_data, data);
+    }
+
+failed:
```

Comment:
Probably this label needs another name, because the code after it will be executed on success and on fail.